### PR TITLE
Add new macros for sqlserver

### PR DESCRIPTION
**Pull request by jkock**

### CODE REVIEW

**Review by alittlesliceoftom — COMMENTED** [review]
@swanderz  any idea why the CI hasn't run on this PR?

**Comment by swanderz** [comment]
> @swanderz any idea why the CI hasn't run on this PR?

there's a setting you have to click for each repo to allow builds to be run on forks. changed it and we're good to go now!

File: macros/dbt_utils/cross_db_utils/hash.sql [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
-    
-    TODO: We can make a small change to dbt-core (https://github.com/fishtown-analytics/dbt/issues/2923)
-    that will make the inheritance of dispatched macros work just like the 
-    inheritance of other adapter objects, and render the following code redundant.
-#}
-{% macro synapse__hash(field) %}
-    {% do return(sqlserver__hash(field)) %}
-{% endmacro %}
+{% macro sqlserver__hash(field) %}
+    CONVERT(VARCHAR(32), HASHBYTES('MD5',  COALESCE(CAST({{field}} AS VARCHAR(MAX) ), '')), 2)
```
[the remark this answers is not in the document]
Comment (swanderz) [reply macros/dbt_utils/cross_db_utils/hash.sql:2]: i'm sure there's a reason, and I'd love to know why!

File: macros/dbt_utils/cross_db_utils/split_part.sql [hunk 2]
```diff
@@ -1,23 +1,19 @@
...
-    TODO: We can make a small change to dbt-core (https://github.com/fishtown-analytics/dbt/issues/2923)
-    that will make the inheritance of dispatched macros work just like the 
-    inheritance of other adapter objects, and render the following code redundant.
-#}
-{% macro synapse__split_part(string_text, delimiter_text, part_number) %}
-    {% do return(sqlserver__split_part(string_text, delimiter_text, part_number)) %}
-{% endmacro %}
+{% macro sqlserver__split_part(string_text, delimiter_text, part_number) %}
+
+    LTRIM(CAST(('<X>'+REPLACE({{ string_text }},{{ delimiter_text }} ,'</X><X>')+'</X>') AS XML).value('(/X)[{{ part_number }}]', 'VARCHAR(128)'))
```
Comment (alittlesliceoftom) [inline macros/dbt_utils/cross_db_utils/split_part.sql:3]: I don't understand this logic so a comment would be lovely, but it passes the Azure SQL tests at least.

File: macros/dbt_utils/cross_db_utils/date_trunc.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{% macro sqlserver__date_trunc(datepart, date) %}
```
Comment (alittlesliceoftom) [inline macros/dbt_utils/cross_db_utils/date_trunc.sql:1]: Fails test with:

  ('22018', '[22018] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Explicit conversion from data type int to date is not allowed. (529) (SQLExecDirectW)')


Though that implies the incoming data is not a date possibly?

Comment (alittlesliceoftom) [reply macros/dbt_utils/cross_db_utils/date_trunc.sql:1]: Ah, so verifying this - the test is that the value is the rounded day. 

select
    cast({{dbt_utils.date_trunc('day', 'updated_at') }} as date) as actual,
    day as expected

from data

the logic here is actually taking the day of the month,  that is in SQL: 
SELECT     datepart(day, '2015-01-01 12:00') 
, returns 1 hence the error. 

The function should return the day if 'day' is supplied, or the first of the month if 'month' is supplied according to the tests. 

Seems like we  could ref last_day.sql, and simply add 1 day and subtract 1 month or many other ways of doing the same.

**Review by alittlesliceoftom — CHANGES_REQUESTED** [review]
Hey @jkock this is excellent, thanks so much for the PR, a big improvement. 

The CI is now running on your PR by some magic, there are 7 tests still failing, but if we can resolve the below comments we can merge as master would fail 15 tests right now (pre-release).

There are 2 issues. 
1 date_trunc is wrong and failing tests (please see my comments and either delete or fix)
2 the hash function has some new complexity so would just love to understand the reasoning before merging.

File: macros/dbt_utils/cross_db_utils/hash.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkock) [reply macros/dbt_utils/cross_db_utils/hash.sql:2]: Good question. I was trying to get the surrogate_key() macro working, and someone suggested this fix for the hash macro. But I also did something on another macro, which probably fixed it. Will revert these changes

File: macros/dbt_utils/cross_db_utils/split_part.sql (hunk 2 above)
Comment (jkock) [reply macros/dbt_utils/cross_db_utils/split_part.sql:7]: Added a link to some more information

File: macros/dbt_utils/cross_db_utils/date_trunc.sql (hunk 3 above)
Comment (jkock) [reply macros/dbt_utils/cross_db_utils/date_trunc.sql:1]: I've changed the behavior

**Comment by swanderz** [comment]
my feeling is to merge these changes in, and slowly chip away at the remaining errors in another PR. I wonder if we can disable some of these models in the interim and add to the README that they aren't yet supported... @jtcohen6 any idea how to do this? perhaps just disable the models?

**Comment by alittlesliceoftom** [comment]
Agreed, excellent progress, thanks for the contribution and the merge!

**Comment by jtcohen6** [comment]
yup, just disable the models! exactly as you did in bc6099f (nice anchor)

**Pull request merged**